commit 95d681115d1962a22a923b3ff1e7d76d56a08b73
Author: Ben Webb <benmwebb@gmail.com>
Date:   Tue Aug 18 11:35:59 2026 -0700

    Add support for Python limited C API
    
    Add a new IMP_PY_LIMITED_API CMake variable
    which, if set, creates Python extensions using
    Python's limited C API, such that they should
    work with multiple Python versions.

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0f61713cc9..40f4b22b92 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -109,6 +109,7 @@ set(SWIG_EXECUTABLE swig CACHE STRING "Swig program")
 set(IMP_DISABLED_MODULES CACHE STRING "A colon-separated list of disabled modules")
 set(IMP_STATIC off CACHE BOOL "Link all binaries and libraries statically; currently only supported on Linux systems with gcc.")
 set(IMP_USE_PYTHON_SOABI off CACHE BOOL "If set, add Python ABI tag to extension modules")
+set(IMP_PY_LIMITED_API "" CACHE STRING "If set, build Python extension modules using the Python limited API at given version (e.g. '3.10') (requires SWIG 4.2 or later)")
 if (${CMAKE_BUILD_TYPE} MATCHES "Debug")
 set(timeout_factor "${timeout_factor} * 2")
 set(IMP_MAX_CHECKS "INTERNAL" CACHE STRING "One of NONE, USAGE, INTERNAL")
diff --git a/tools/build/cmake_templates/Module.cmake b/tools/build/cmake_templates/Module.cmake
index 61b8672e62..808345c9d5 100644
--- a/tools/build/cmake_templates/Module.cmake
+++ b/tools/build/cmake_templates/Module.cmake
@@ -22,7 +22,8 @@ if(${status} EQUAL 0)
     COMMAND ${PYTHON_EXECUTABLE} %(tools_dir)sbuild/setup_swig_wrappers.py
     %(build_dir)s --module=%(name)s
     --datapath=${IMP_DATAPATH}
-    --source=${CMAKE_SOURCE_DIR})
+    --source=${CMAKE_SOURCE_DIR}
+    --py_limited_api=${IMP_PY_LIMITED_API})
 
   # for warning control
   add_definitions(-DIMP%(NAME)s_COMPILATION)
diff --git a/tools/build/cmake_templates/ModuleSwig.cmake b/tools/build/cmake_templates/ModuleSwig.cmake
index 56f67de853..90b2774c45 100644
--- a/tools/build/cmake_templates/ModuleSwig.cmake
+++ b/tools/build/cmake_templates/ModuleSwig.cmake
@@ -69,8 +69,16 @@ if(WIN32 AND NOT CYGWIN)
   set_target_properties(IMP.%(name)s-python PROPERTIES SUFFIX ".pyd")
 endif()
 
-if(IMP_USE_PYTHON_SOABI AND NOT WIN32 AND DEFINED PYTHON_SOABI AND NOT PYTHON_SOABI STREQUAL "")
-  set_target_properties(IMP.%(name)s-python PROPERTIES SUFFIX ".${PYTHON_SOABI}.so")
+if(IMP_PY_LIMITED_API)
+  if(WIN32 AND NOT CYGWIN)
+    set_target_properties(IMP.%(name)s-python PROPERTIES SUFFIX ".abi3.pyd")
+  else()
+    set_target_properties(IMP.%(name)s-python PROPERTIES SUFFIX ".abi3.so")
+  endif()
+else()
+  if(IMP_USE_PYTHON_SOABI AND NOT WIN32 AND DEFINED PYTHON_SOABI AND NOT PYTHON_SOABI STREQUAL "")
+    set_target_properties(IMP.%(name)s-python PROPERTIES SUFFIX ".${PYTHON_SOABI}.so")
+  endif()
 endif()
 
 target_link_libraries(IMP.%(name)s-python
diff --git a/tools/build/setup_swig_wrappers.py b/tools/build/setup_swig_wrappers.py
index f83e2466d1..fc7ad98460 100755
--- a/tools/build/setup_swig_wrappers.py
+++ b/tools/build/setup_swig_wrappers.py
@@ -6,6 +6,7 @@
 import os.path
 import datetime
 import tools
+import re
 from argparse import ArgumentParser
 
 
@@ -40,7 +41,7 @@ def write_module_swig(m, contents, skip_import=False):
         contents.append("%%import \"IMP_%(module)s.i\"" % {"module": m.name})
 
 
-def build_wrapper(module, finder, sorted, target, source):
+def build_wrapper(module, finder, sorted, target, source, limapi):
     if not module.configured.ok or module.python_only:
         return
     contents = []
@@ -57,18 +58,27 @@ def build_wrapper(module, finder, sorted, target, source):
 # Copyright 2007-%d IMP Inventors. All rights reserved.
 
 %%}
+""" % (swig_module_name, year))  # noqa: E501
+
+    if limapi:
+        contents.append("/* Work with any Python version >= %d.%d */"
+                        % limapi)
+        contents.append("%%begin %%{\n"
+                        "#define Py_LIMITED_API 0x%02x%02x0000\n%%}"
+                        % limapi)
 
+    contents.append("""
 /* '#' formats in parsing or building Python values
    (e.g. in PyObject_CallFunction) use Py_ssize_t, not int, for lengths
    in Python >= 2.5 */
-%%begin %%{
+%begin %{
 #define PY_SSIZE_T_CLEAN
-%%}
+%}
 
 // Warning 314: 'lambda' is a python keyword, renaming to '_lambda'
-%%warnfilter(321,302,314);
+%warnfilter(321,302,314);
 
-%%{
+%{
 #include <boost/version.hpp>
 #include <boost/exception/all.hpp>
 
@@ -84,8 +94,8 @@ extern "C"
 // suppress warning
 SWIGEXPORT
 PyObject* SWIG_init();
-%%}
-""" % (swig_module_name, year))  # noqa: E501
+%}
+""")  # noqa: E501
     # some of the typemap code ends up before this is swig sees the
     # typemaps first
     all_deps = [x for x in finder.get_dependent_modules([module])
@@ -154,10 +164,22 @@ parser.add_argument("-s", "--source",
                     dest="source", help="Where to find IMP source.")
 parser.add_argument("-m", "--module",
                     dest="module", default="", help="Only run on one module.")
+parser.add_argument(
+    "--py_limited_api", dest="limapi", default="",
+    help="If set, build Python extension modules using the Python limited API "
+         "at given version (e.g. '3.10').")
 
 
 def main():
     args = parser.parse_args()
+    if args.limapi:
+        pyver = re.match(r'(\d+)\.(\d+)$', args.limapi)
+        if not pyver:
+            raise ValueError("Cannot parse Python limited API (%r) as "
+                             "an X.Y version number" % args.limapi)
+        limapi = (int(pyver.group(1)), int(pyver.group(2)))
+    else:
+        limapi = None
     sorted_order = tools.get_sorted_order()
 
     if args.module != "":
@@ -167,7 +189,7 @@ def main():
         module = mf[args.module]
         build_wrapper(module, mf, sorted_order,
                       os.path.join("swig", "IMP_" + module.name + ".i"),
-                      args.source)
+                      args.source, limapi)
 
 
 if __name__ == '__main__':
